File: python/benchmarks/benchmark.py
import sys
import os
import argparse
import time
import numpy as np
import cv2
import json

# Add project root to sys.path
base_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../'))
if base_path not in sys.path:
    sys.path.insert(0, os.path.join(base_path, 'python'))

from modules.model import RFDETRModel

def run_benchmark(model_path, data_path, device, num_iterations=100, warmup_iterations=10, sleep_per_image=0.0, verbose=False):
    print(f"Initializing model: {model_path} on {device}")
    model = RFDETRModel(model_path, device=device)
    
    # Identify if data_path is directory, video, or image
    is_dir = os.path.isdir(data_path)
    is_video = data_path.lower().endswith(('.mp4', '.avi', '.mov', '.mkv'))
    
    images = []
    if is_dir:
        for f in os.listdir(data_path):
            if f.lower().endswith(('.jpg', '.jpeg', '.png', '.bmp')):
                images.append(cv2.imread(os.path.join(data_path, f)))
        if not images:
            print(f"Error: No images found in {data_path}")
            sys.exit(1)
        print(f"Loaded {len(images)} images from directory.")
    elif is_video:
        cap = cv2.VideoCapture(data_path)
        if not cap.isOpened():
            print(f"Error: Could not open video {data_path}")
            sys.exit(1)
        # Read a few frames for warmup and benchmarking
        count = 0
        while count < num_iterations + warmup_iterations:
            ret, frame = cap.read()
            if not ret: break
            images.append(frame)
            count += 1
        cap.release()
        print(f"Loaded {len(images)} frames from video.")
    else:
        img = cv2.imread(data_path)
        if img is None:
            print(f"Error: Could not load image {data_path}")
            sys.exit(1)
        images = [img]

    print(f"Running {warmup_iterations} warmup iterations...")
    for i in range(warmup_iterations):
        model.predict(images[i % len(images)])

    print(f"Running {num_iterations} benchmark iterations...")
    
    pre_times = []
    ort_times = []
    post_times = []
    total_times = []

    for i in range(num_iterations):
        if sleep_per_image > 0:
            time.sleep(sleep_per_image)
        img_idx = i % len(images)
        _, _, _, _, timings = model.predict(images[img_idx])
        
        pre = timings['preprocess']
        ort = timings['ort_run']
        post = timings['postprocess']
        total = pre + ort + post
        
        pre_times.append(pre)
        ort_times.append(ort)
        post_times.append(post)
        total_times.append(total)
        
        if verbose:
            print(f"Iteration {i+1:3d}: Pre: {pre:6.2f}ms, ORT: {ort:6.2f}ms, Post: {post:6.2f}ms, Total: {total:6.2f}ms")
        elif (i + 1) % 10 == 0:
            print(f"Iteration {i+1}/{num_iterations}")

    results = {
        "implementation": "Python",
        "device": device,
        "num_iterations": num_iterations,
        "data_source": data_path,
        "metrics": {
            "preprocessing": {
                "mean": np.mean(pre_times),
                "median": np.median(pre_times),
                "std": np.std(pre_times),
                "min": np.min(pre_times),
                "max": np.max(pre_times)
            },
            "ort_run": {
                "mean": np.mean(ort_times),
                "median": np.median(ort_times),
                "std": np.std(ort_times),
                "min": np.min(ort_times),
                "max": np.max(ort_times)
            },
            "postprocessing": {
                "mean": np.mean(post_times),
                "median": np.median(post_times),
                "std": np.std(post_times),
                "min": np.min(post_times),
                "max": np.max(post_times)
            },
            "total_processing": {
                "mean": np.mean(total_times),
                "median": np.median(total_times),
                "std": np.std(total_times),
                "min": np.min(total_times),
                "max": np.max(total_times),
                "fps": 1000.0 / np.median(total_times)
            }
        },
        "iterations": [
            {
                "index": i + 1,
                "preprocessing": pre_times[i],
                "ort_run": ort_times[i],
                "postprocessing": post_times[i],
                "total": total_times[i]
            } for i in range(num_iterations)
        ]
    }
    
    return results

def main():
    parser = argparse.ArgumentParser(description="Python Benchmark for RF-DETR ONNX")
    parser.add_argument("--model", required=True, type=str, help="Path to ONNX model")
    parser.add_argument("--input", required=True, type=str, help="Path to input image, directory, or video")
    parser.add_argument("--device", default="gpu", choices=["gpu", "cpu"], help="Device to use")
    parser.add_argument("--iterations", default=100, type=int, help="Number of iterations")
    parser.add_argument("--sleep-per-image", default=0.0, type=float, help="Seconds to sleep between images (default: 0.0)")
    parser.add_argument("--verbose", action="store_true", help="Enable per-iteration logging")
    parser.add_argument("--output", default="benchmark_python.json", type=str, help="Output JSON file")
    
    args = parser.parse_args()
    
    results = run_benchmark(args.model, args.input, args.device, args.iterations, 
                            sleep_per_image=args.sleep_per_image, verbose=args.verbose)
    
    with open(args.output, "w") as f:
        json.dump(results, f, indent=4)
    
    print(f"Results saved to {args.output}")

if __name__ == "__main__":
    main()

File: cpp/benchmarks/benchmark.cpp
#include "rfdetr_model.hpp"
#include <algorithm>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <numeric>
#include <opencv2/opencv.hpp>
#include <thread>
#include <vector>

namespace fs = std::filesystem;

struct Stats {
  float mean;
  float median;
  float std;
  float min;
  float max;
};

Stats calculateStats(std::vector<float> times) {
  if (times.empty())
    return {0, 0, 0, 0, 0};

  float sum = std::accumulate(times.begin(), times.end(), 0.0f);
  float mean = sum / times.size();

  float sq_sum =
      std::inner_product(times.begin(), times.end(), times.begin(), 0.0f);
  float std = std::sqrt(sq_sum / times.size() - mean * mean);

  auto [min_it, max_it] = std::minmax_element(times.begin(), times.end());
  float min_val = *min_it;
  float max_val = *max_it;

  // Calculate median
  std::sort(times.begin(), times.end());
  float median;
  size_t size = times.size();
  if (size % 2 == 0) {
    median = (times[size / 2 - 1] + times[size / 2]) / 2;
  } else {
    median = times[size / 2];
  }

  return {mean, median, std, min_val, max_val};
}

int main(int argc, char **argv) {
  if (argc < 4) {
    std::cerr << "Usage: " << argv[0]
              << " <model_path> <input_path> <device> [iterations]"
              << std::endl;
    return 1;
  }

  std::string modelPath = argv[1];
  std::string inputPath = argv[2];
  std::string device = argv[3];
  int numIterations = (argc > 4) ? std::stoi(argv[4]) : 100;
  float sleepPerImage = (argc > 5) ? std::stof(argv[5]) : 0.0f;
  bool verbose = (argc > 6 && std::string(argv[6]) == "verbose");
  int warmupIterations = 10;

  std::cout << "Initializing model: " << modelPath << " on " << device
            << std::endl;
  rfdetr::RFDETRModel model(modelPath, device);

  std::vector<cv::Mat> images;

  // Check if input is a directory, video, or image
  if (fs::is_directory(inputPath)) {
    for (const auto &entry : fs::directory_iterator(inputPath)) {
      std::string ext = entry.path().extension().string();
      std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
      if (ext == ".jpg" || ext == ".jpeg" || ext == ".png") {
        cv::Mat img = cv::imread(entry.path().string());
        if (!img.empty())
          images.push_back(img);
      }
    }
    if (images.empty()) {
      std::cerr << "Error: No images found in directory " << inputPath
                << std::endl;
      return 1;
    }
    std::cout << "Loaded " << images.size() << " images from directory."
              << std::endl;
  } else if (inputPath.find(".mp4") != std::string::npos ||
             inputPath.find(".avi") != std::string::npos) {
    cv::VideoCapture cap(inputPath);
    if (!cap.isOpened()) {
      std::cerr << "Error: Could not open video " << inputPath << std::endl;
      return 1;
    }
    cv::Mat frame;
    int count = 0;
    while (count < numIterations + warmupIterations && cap.read(frame)) {
      images.push_back(frame.clone());
      count++;
    }
    cap.release();
    std::cout << "Loaded " << images.size() << " frames from video."
              << std::endl;
  } else {
    cv::Mat img = cv::imread(inputPath);
    if (img.empty()) {
      std::cerr << "Error: Could not load image " << inputPath << std::endl;
      return 1;
    }
    images.push_back(img);
  }

  std::cout << "Running " << warmupIterations << " warmup iterations..."
            << std::endl;
  for (int i = 0; i < warmupIterations; ++i) {
    std::vector<rfdetr::Detection> dets;
    rfdetr::Timings timings;
    model.predict(images[i % images.size()], dets, timings);
  }

  std::cout << "Running " << numIterations << " benchmark iterations..."
            << std::endl;
  std::vector<float> pre_times, ort_times, post_times, total_processing_times;

  for (int i = 0; i < numIterations; ++i) {
    if (sleepPerImage > 0) {
      std::this_thread::sleep_for(
          std::chrono::milliseconds(static_cast<int>(sleepPerImage * 1000)));
    }
    std::vector<rfdetr::Detection> dets;
    rfdetr::Timings timings;
    model.predict(images[i % images.size()], dets, timings);

    float total = timings.preprocess + timings.ort_run + timings.postprocess;
    pre_times.push_back(timings.preprocess);
    ort_times.push_back(timings.ort_run);
    post_times.push_back(timings.postprocess);
    total_processing_times.push_back(total);

    if (verbose) {
      printf("Iteration %3d: Pre: %6.2fms, ORT: %6.2fms, Post: %6.2fms, Total: "
             "%6.2fms\n",
             i + 1, timings.preprocess, timings.ort_run, timings.postprocess,
             total);
    } else if ((i + 1) % 10 == 0) {
      std::cout << "Iteration " << (i + 1) << "/" << numIterations << std::endl;
    }
  }

  auto pre_stats = calculateStats(pre_times);
  auto ort_stats = calculateStats(ort_times);
  auto post_stats = calculateStats(post_times);
  auto total_stats = calculateStats(total_processing_times);

  std::string outputFilename = "benchmark_cpp_" + device + ".json";
  std::ofstream ofs(outputFilename);
  ofs << "{\n"
      << "  \"implementation\": \"C++\",\n"
      << "  \"device\": \"" << device << "\",\n"
      << "  \"num_iterations\": " << numIterations << ",\n"
      << "  \"data_source\": \"" << inputPath << "\",\n"
      << "  \"metrics\": {\n"
      << "    \"preprocessing\": {\"mean\": " << pre_stats.mean
      << ", \"median\": " << pre_stats.median << ", \"std\": " << pre_stats.std
      << ", \"min\": " << pre_stats.min << ", \"max\": " << pre_stats.max
      << "},\n"
      << "    \"ort_run\": {\"mean\": " << ort_stats.mean
      << ", \"median\": " << ort_stats.median << ", \"std\": " << ort_stats.std
      << ", \"min\": " << ort_stats.min << ", \"max\": " << ort_stats.max
      << "},\n"
      << "    \"postprocessing\": {\"mean\": " << post_stats.mean
      << ", \"median\": " << post_stats.median
      << ", \"std\": " << post_stats.std << ", \"min\": " << post_stats.min
      << ", \"max\": " << post_stats.max << "},\n"
      << "    \"total_processing\": {\"mean\": " << total_stats.mean
      << ", \"median\": " << total_stats.median
      << ", \"std\": " << total_stats.std << ", \"min\": " << total_stats.min
      << ", \"max\": " << total_stats.max
      << ", \"fps\": " << 1000.0 / total_stats.median << "}\n"
      << "  },\n"
      << "  \"iterations\": [\n";

  for (int i = 0; i < numIterations; ++i) {
    ofs << "    {\n"
        << "      \"index\": " << i + 1 << ",\n"
        << "      \"preprocessing\": " << pre_times[i] << ",\n"
        << "      \"ort_run\": " << ort_times[i] << ",\n"
        << "      \"postprocessing\": " << post_times[i] << ",\n"
        << "      \"total\": " << total_processing_times[i] << "\n"
        << "    }" << (i == numIterations - 1 ? "" : ",") << "\n";
  }

  ofs << "  ]\n"
      << "}\n";

  std::cout << "Results saved to " << outputFilename << std::endl;

  return 0;
}
